cli/azd/cmd/cobra_builder.go

```diff
@@ -0,0 +1,221 @@
+}
+
+// Creates a new instance of the Cobra builder
+func NewCobraBuilder(container *ioc.NestedContainer) *CobraBuilder {
+	return &CobraBuilder{
+		container: container,
+		runner:    middleware.NewMiddlewareRunner(container),
+	}
+}
+
+// Builds a cobra Command for the specified action descriptor
+func (cb *CobraBuilder) BuildCommand(descriptor *actions.ActionDescriptor) (*cobra.Command, error) {
+	cmd := descriptor.Options.Command
+	if cmd.Use == "" {
+		cmd.Use = descriptor.Name
+	}
+
+	// Build the full command tree
+	for _, childDescriptor := range descriptor.Children() {
+		childCmd, err := cb.BuildCommand(childDescriptor)
+		if err != nil {
+			return nil, err
+		}
+
+		cmd.AddCommand(childCmd)
+	}
+
+	// Bind root command after command tree has been established
+	// This ensures the command path is ready and consistent across all nested commands
+	if descriptor.Parent() == nil {
+		if err := cb.bindCommand(cmd, descriptor); err != nil {
+			return nil, err
+		}
+	}
+
+	cb.configureActionResolver(cmd, descriptor)
+
+	return cmd, nil
+}
+
+// Configures the cobra command 'RunE' function to running the composed middleware and action for the
+// current action descriptor
+func (cb *CobraBuilder) configureActionResolver(cmd *cobra.Command, descriptor *actions.ActionDescriptor) {
+	// Only bind command to action if an action resolver had been defined
+	// and when a RunE hasn't already been set
+	if descriptor.Options.ActionResolver == nil || cmd.RunE != nil {
+		return
+	}
+
+	cmd.RunE = func(cmd *cobra.Command, args []string) error {
+		ctx := cmd.Context()
+		ctx = tools.WithInstalledCheckCache(ctx)
+
+		// Registers the following to enable injection into actions that require them
+		ioc.RegisterInstance(ioc.Global, cb.runner)
+		ioc.RegisterInstance(ioc.Global, ctx)
+		ioc.RegisterInstance(ioc.Global, cmd)
+		ioc.RegisterInstance(ioc.Global, args)
+
+		if err := cb.registerMiddleware(descriptor); err != nil {
+			return err
+		}
+
+		actionName := createActionName(cmd)
+		var action actions.Action
+		if err := cb.container.ResolveNamed(actionName, &action); err != nil {
+			return fmt.Errorf(
+				//nolint:lll
+				"failed resolving action '%s'. Ensure the ActionResolver is a valid go function that returns an `actions.Action` interface, %w",
+				actionName,
+				err,
+			)
+		}
+
+		runOptions := &middleware.Options{
+			Name:    cmd.CommandPath(),
+			Aliases: cmd.Aliases,
+		}
+
+		// Run the middleware chain with action
+		log.Printf("Resolved action '%s'\n", actionName)
+		_, err := cb.runner.RunAction(ctx, runOptions, action)
+
+		return err
+	}
+}
+
+// Binds the intersection of cobra command options and action descriptor options
+func (cb *CobraBuilder) bindCommand(cmd *cobra.Command, descriptor *actions.ActionDescriptor) error {
+	actionName := createActionName(cmd)
+
+	// Automatically adds a consistent help flag
+	cmd.Flags().BoolP("help", "h", false, fmt.Sprintf("Gets help for %s.", descriptor.Name))
```

Comment:
Should this be using `cmd.Use` instead to always be logically correct? An `ActionDescriptor` can be constructed with an `ActionDescriptorOptions` with a `cmd.Use`, and that would **not** be overridden if already defined.

Reply:
Good catch! Updating to use `cmd.Name()` as we were using before that returns the first word of the `cmd.Use` param.